### PR TITLE
sys/xtimer: code simplification (removing long_list, using xtimer_remove, removing redundant expiry check)

### PR DESCRIPTION
**Pull request by Hyungsin**
### Contributions
@MichelRottleuthner Remove redundant part of xtimer_core.c (e.g., long timer list). I put some comments on the code.

### CODE REVIEW

File: sys/xtimer/xtimer_core.c [hunk 1]
```diff
@@ -281,32 +240,13 @@ static inline void _update_short_timers(uint64_t *now)
...
-        if (timer_list_head->offset < elapsed ||
-            timer_list_head->offset - elapsed < XTIMER_ISR_BACKOFF) {
-            goto update;
-        }
-        else {
-            timer_list_head->offset -= elapsed;
-            timer_list_head->start_time = (uint32_t)now;
-            timer_list_head->long_start_time = (uint32_t)(now >> 32);
-        }
-    }
```
Comment (Hyungsin) [inline sys/xtimer/xtimer_core.c]: _update_timers already did timer expiry check and it does not have to be re-done here.

File: sys/xtimer/xtimer_core.c [hunk 2]
```diff
@@ -201,66 +194,32 @@ void xtimer_remove(xtimer_t *timer)
...
             if (timer->offset > elapsed) {
                 while(_xtimer_now() - timer->start_time < timer->offset) {}
             }
-            /* advance list */
-            timer_list_head = timer->next;
-            /* make sure timer is recognized as being already fired */
-            timer->offset = 0;
-            timer->start_time = 0;
-            timer->long_start_time = 0;
-            timer->next = NULL;
```
Comment (Hyungsin) [inline sys/xtimer/xtimer_core.c]: A few lines above are actually what **xtimer_remove()** can do.

File: sys/xtimer/xtimer_core.c [hunk 3]
```diff
@@ -40,13 +40,11 @@ static volatile int _in_handler = 0;
 volatile uint64_t _xtimer_current_time = 0;
 
 static xtimer_t *timer_list_head = NULL;
-static xtimer_t *long_list_head = NULL;
```
Comment (Hyungsin) [inline sys/xtimer/xtimer_core.c]: does not need to manage a separate long_list. One timer list is actually enough.

**Comment by fjmolinas** [comment]
@Hyungsin is this only cleanup?

**Comment by fjmolinas** [comment]
@Hyungsin also in general all this comments you make in the PR are great but it would be even better to have them in the commit message.

**Comment by Hyungsin** [comment]
> @Hyungsin is this only cleanup?

Yes. I think so.

**Comment by Hyungsin** [comment]
> @Hyungsin also in general all this comments you make in the PR are great but it would be even better to have them in the commit message.

Added details on commit msg.

**Comment by miri64** [comment]
Huh? Nobody reviewing this in 7 month? @MichelRottleuthner can you have a look?

**Comment by fjmolinas** [comment]
ping @MichelRottleuthner!

**Comment by MichelRottleuthner** [comment]
The changes make sense but as with all timer related things this needs solid testing to not break something. Sadly I cannot allocate time for this currently.
@pokgak did you consider this for your measurements? Would be interesting to see if/how it affects performance as the measurements indicate there is some headroom that is probably related to this list handling.

**Comment by fjmolinas** [comment]
> The changes make sense but as with all timer related things this needs solid testing to not break something. Sadly I cannot allocate time for this currently.
> @pokgak did you consider this for your measurements? Would be interesting to see if/how it affects performance as the measurements indicate there is some headroom that is probably related to this list handling.

If testing is what is missing do you mind adding the other review flags :)

**Comment by pokgak** [comment]
I ran my benchmark on current master (bfb8fc52b) and on this PR after a rebase to the same master. The benchmark test is largely based on the `tests/bench_timers` with some extra measurements added. The code can be found [here](https://github.com/pokgak/RobotFW-tests/blob/ba/timer-benchmarks/tests/xtimer_benchmarks/main.c).

There are some conflict when rebasing, so I just accept all incoming changes. The diff between the rebased PR and master (from the rebased branch `git diff master -- sys/xtimer/xtimer_core.c` can be found [here](https://gist.github.com/pokgak/7edfc5e73488561c78b78260da51c22d). Results for the master branch can be found [here](https://pokgak.github.io/ba-plotscripts/pr13103/master/) and for this PR after rebase [here](https://pokgak.github.io/ba-plotscripts/pr13103/pr_rebased/). You can open the links in separate browser windows for easier side-by-side comparison. Note that the y-axis is not the same for the graph.

One my test (sleep jitter) timed out on the rebased PR. The test have a periodic timer triggering every 100ms with a number of timer triggering randomly in the background. At 10 background timer, it is still okay but at 20 the tests takes too long (more than 45 seconds) and timed out. I didn't investigate further into as to what causes this.

**Comment by miri64** [comment]
Please also check for the bug fixed in https://github.com/RIOT-OS/RIOT/pull/13850. The test introduced there should cover that, but IMHO it doesn't hurt (except time spent) to also re-test the bug described there under "Issues/PRs references".

**Pull request closed without merging**